tests/cpp/test_multidevice_lower_communication.cpp

```diff
@@ -202,6 +202,38 @@ TEST_F(LowerCollectiveTest, Allgather) {
   EXPECT_TRUE(at::equal(out_tensor, unsharded_tensor));
 }
 
+TEST_F(LowerCollectiveTest, Allgather_LoopSplit) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  const auto num_devices = communicator_->size();
+  auto mesh = DeviceMesh::createForNumDevices(num_devices);
+
+  TensorView* in = makeContigTensor(1);
+  in->setDeviceMesh(mesh);
+  TensorView* out = set(in);
+  fusion->addInput(in);
+  fusion->addOutput(out);
+
+  in->split(0, num_devices, /*inner_split=*/false);
+  in->axis(0)->parallelize(ParallelType::DIDx);
+  in->setAllocationDomain(in->getLoopDomain(), true);
+
+  out->split(0, num_devices, /*inner_split=*/false);
+  out->setAllocationDomain(out->getLoopDomain(), true);
+
+  at::Tensor unsharded_tensor =
+      at::randn({num_devices * kTensorSize}, at::kFloat);
+  at::Tensor in_tensor =
+      shardTensor(unsharded_tensor, in).to(communicator_->device());
+
```

Comment:
```suggestion

std::vector<int64_t> ref_int_tensor_shape = {kTensorSize};
EXPECT_EQ(in_tensor.sizes(), ref_int_tensor_shape);

```

Reply:
I don't see how shardTensor is correct here if it never play the split backwards... But I might be missing something.

Reply:
Re the suggested change: I manually checked the shape is as expected. I added some extra unit tests for shardTensor alone, so we don't have to verify it here.

Reply:
I made a couple of changes to address the problems I said in https://github.com/NVIDIA/Fuser/pull/3284#discussion_r1868843000. 
1. https://github.com/NVIDIA/Fuser/pull/3284/commits/7cf238402fdbd0656f26577088128418e7433f6d. It's an overkill but will probably be OK for quite some time. I had a hard time finding a concrete use case that has to mix DID and host ID within one logical dimension. I agree that to properly support inner splits we'll need to "replay the split backwards". It's not a trivial change anyhow so I'll postpone it to a separate PR. 
2. I wrote https://github.com/NVIDIA/Fuser/pull/3531 to harden runtime checks for allgather and added to this PR one more allgather test (Allgather_LoopSplit_Noncontiguous). These extra checks will fire when we trigger some most common limitations before properly fixing ReorderShardedAxisPass, which will take several decent-size PRs.

Reply:
> I had a hard time finding a concrete use case that has to mix DID and host ID within one logical dimension. 

In fact, there's https://github.com/NVIDIA/Fuser/blob/64bc5604f5b14a4052bb579426133c028662a934/tests/cpp/test_multidevice_overlap.cpp#L681. So I'll try to file a feature request after this PR.